code/modules/shieldgen/shield_gen.dm

```diff
@@ -213,23 +213,35 @@
 
 	var/turf/T
 
-	for (var/x_offset = -field_radius; x_offset <= field_radius; x_offset++)
-		T = locate(gen_turf.x + x_offset, gen_turf.y - field_radius, gen_turf.z)
-		if (T)
-			. += T
-
-		T = locate(gen_turf.x + x_offset, gen_turf.y + field_radius, gen_turf.z)
-		if (T)
-			. += T
-
-	for (var/y_offset = -field_radius+1; y_offset < field_radius; y_offset++)
-		T = locate(gen_turf.x - field_radius, gen_turf.y + y_offset, gen_turf.z)
-		if (T)
-			. += T
-
-		T = locate(gen_turf.x + field_radius, gen_turf.y + y_offset, gen_turf.z)
-		if (T)
-			. += T
+	var/connected_levels = list(gen_turf)
+	if(multiz)
+		var/turf/above = getzabove(src)
+		var/turf/below = getzbelow(src)
+		if(above)
+			for(var/turf/z in above)
+				connected_levels += z
+		if(below)
+			for(var/turf/z in below)
+				connected_levels += z
+
+	for(var/turf/z_level in connected_levels)
+		for (var/x_offset = -field_radius; x_offset <= field_radius; x_offset++)
+			T = locate(z_level.x + x_offset, z_level.y - field_radius, z_level.z)
+			if (T)
+				. += T
+
+			T = locate(z_level.x + x_offset, z_level.y + field_radius, z_level.z)
+			if (T)
+				. += T
+
+		for (var/y_offset = -field_radius+1; y_offset < field_radius; y_offset++)
+			T = locate(z_level.x - field_radius, z_level.y + y_offset, z_level.z)
+			if (T)
+				. += T
+
+			T = locate(z_level.x + field_radius, z_level.y + y_offset, z_level.z)
+			if (T)
+				. += T
```

Comment:
You could use `block()` here, might be faster / simpler? To get the perimeter of the square in four sections without loops.

Also you should maybe cover the case where we're creating the shield at the edge of the map. Unless you want the shield to be missing one side in that case?